artifact/tar/platform/libarchive/wrapper.cpp

```diff
@@ -0,0 +1,142 @@
+// the size of the data blocks returned.
+//
+// - On EOF return 0.
+// - On error return -1.
+ssize_t reader_callback(archive *archive, void *in_reader_container, const void **buff) {
+	ReaderContainer *p_reader_container = static_cast<ReaderContainer *>(in_reader_container);
+
+	auto ret = p_reader_container->reader_.Read(
+		p_reader_container->buff_.begin(), p_reader_container->buff_.end());
+	if (!ret) {
+		log::Error("Failed to read from the archive stream: Error: " + ret.error().message);
+		return -1;
+	}
+
+	size_t bytes_read {ret.value()};
+	*buff = &p_reader_container->buff_.data()[0];
+
+	return bytes_read;
+};
+
+Error Handle::Init() {
+	int r;
+#ifdef MENDER_ARTIFACT_GZIP_COMPRESSION
+	r = archive_read_support_filter_gzip(archive_.get());
+	if (r != ARCHIVE_OK) {
+		return MakeError(error::GenericError, "Gzip compression is not supported on this platform");
+	}
+#endif // MENDER_ARTIFACT_GZIP_COMPRESSION
+#ifdef MENDER_ARTIFACT_LZMA_COMPRESSION
+	r = archive_read_support_filter_xz(archive_.get());
+	if (r != ARCHIVE_OK) {
+		return MakeError(error::GenericError, "xz compression is not supported on this platform");
+	}
+#endif // MENDER_ARTIFACT_LZMA_COMPRESSION
+#ifdef MENDER_ARTIFACT_ZSTD_COMPRESSION
+	r = archive_read_support_filter_zstd(archive_.get());
+	if (r != ARCHIVE_OK) {
+		return MakeError(error::GenericError, "xz compression is not supported on this platform");
+	}
+#endif // MENDER_ARTIFACT_ZSTD_COMPRESSION
+	r = archive_read_support_format_tar(archive_.get());
+	if (r != ARCHIVE_OK) {
+		return MakeError(error::GenericError, "the tar format is not supported on this platform");
+	}
+	r = archive_read_open(archive_.get(), &reader_container_, nullptr, reader_callback, nullptr);
+	if (r != ARCHIVE_OK) {
+		return MakeError(
+			error::GenericError,
+			"Failed to initalize the 'libarchive' C bindings. LibArchive error message: '"
+				+ string(archive_error_string(archive_.get()))
+				+ "' error code: " + std::to_string(archive_errno(archive_.get())));
+	}
+	this->initalized_ = true;
+	return error::NoError;
+}
+
+int FreeLibArchiveHandle(archive *a) {
+	int r {0};
+	if (a != nullptr) {
+		r = archive_read_free(a);
+		if (r != ARCHIVE_OK) {
+			log::Error("Failed to free the resources from the Archive");
+		}
+	}
+	return r;
+}
+
+Handle::Handle(io::Reader &reader) :
+	archive_(archive_read_new(), FreeLibArchiveHandle),
+	reader_container_ {reader, libarchive_read_buffer_size} {
+	auto err = Init();
+	if (err) {
+		log::Error("Failed to initialize the Archive handle" + err.message);
+	}
+}
+
+
+ExpectedSize Handle::Read(vector<uint8_t>::iterator start, vector<uint8_t>::iterator end) {
+	if (!initalized_) {
+		return common::error::MakeError(
+			common::error::GenericError,
+			"Unable to read from a tar reader which is not initialized properly");
+	}
+	size_t iterator_size {static_cast<size_t>(end - start)};
+	ssize_t read_bytes {archive_read_data(archive_.get(), &start[0], iterator_size)};
```

Comment:
I'd use good old assignment for these, but YMMV.

Reply:
https://google.github.io/styleguide/cppguide.html#Variable_and_Array_Initialization

Seems they allow both